src/scala/org/pantsbuild/zinc/analysis/AnalysisMap.scala

```diff
@@ -65,13 +66,22 @@ class AnalysisMap private[AnalysisMap] (
      * Can remove after the sbt jar output patch lands.
      */
     def definesClass(classpathEntry: File): DefinesClass = {
-      getAnalysis(classpathEntry).map { analysis =>
-        // strongly hold the classNames, and transform them to ensure that they are unlinked from
-        // the remainder of the analysis
-        val classNames = analysis.asInstanceOf[Analysis].relations.srcProd.reverseMap.keys.toList.toSet.map(
-          (f: File) => filePathToClassName(f))
-        new ClassNamesDefinesClass(classNames)
-      }.getOrElse {
+      // If we have analysis with a valid Compilation, use the classnames it refers to.
+      val analysisDefinesClass =
```

Comment:
Could be worth annotating this with a type, it took me a while to determine that this for comprehension was single-element (an `Option`) and not an iterator.